src/sources/amqp.rs

```diff
@@ -0,0 +1,565 @@
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        true
+    }
+}
+
+#[derive(Debug)]
+struct FinalizerEntry {
+    acker: Acker,
+}
+
+impl From<Delivery> for FinalizerEntry {
+    fn from(delivery: Delivery) -> Self {
+        Self {
+            acker: delivery.acker,
+        }
+    }
+}
+
+pub(crate) async fn amqp_source(
+    config: &AMQPSourceConfig,
+    shutdown: ShutdownSignal,
+    out: SourceSender,
+    log_namespace: LogNamespace,
+    acknowledgements: bool,
+) -> crate::Result<super::Source> {
+    let config = config.clone();
+    let (_conn, channel) = config
+        .connection
+        .connect()
+        .await
+        .map_err(|source| BuildError::AMQPCreateError { source })?;
+
+    Ok(Box::pin(run_amqp_source(
+        config,
+        shutdown,
+        out,
+        channel,
+        log_namespace,
+        acknowledgements,
+    )))
+}
+
+struct Keys<'a> {
+    routing_key: &'a str,
+    routing: &'a str,
+    exchange_key: &'a str,
+    exchange: &'a str,
+    offset_key: &'a str,
+    delivery_tag: i64,
+}
+
+/// Populates the decoded event with extra metadata.
+fn populate_event(
+    event: &mut Event,
+    timestamp: chrono::DateTime<Utc>,
+    keys: &Keys<'_>,
+    log_namespace: LogNamespace,
+) {
+    let log = event.as_mut_log();
+
+    log_namespace.insert_vector_metadata(
+        log,
+        log_schema().timestamp_key(),
+        "ingest_timestamp",
+        timestamp,
+    );
+
+    log_namespace.insert_vector_metadata(
+        log,
+        log_schema().source_type_key(),
+        "source_type",
+        "amqp",
+    );
+
+    log_namespace.insert_source_metadata(
+        "amqp",
+        log,
+        keys.routing_key,
+        "routing",
+        keys.routing.to_string(),
+    );
+
+    log_namespace.insert_source_metadata(
+        "amqp",
+        log,
+        keys.exchange_key,
+        "exchange",
+        keys.exchange.to_string(),
+    );
+
+    log_namespace.insert_source_metadata("amqp", log, keys.offset_key, "offset", keys.delivery_tag);
+}
+
+/// Runs the AMQP source involving the main loop pulling data from the server.
+async fn run_amqp_source(
+    config: AMQPSourceConfig,
+    shutdown: ShutdownSignal,
+    mut out: SourceSender,
+    channel: Channel,
+    log_namespace: LogNamespace,
+    acknowledgements: bool,
+) -> Result<(), ()> {
+    let (finalizer, mut ack_stream) =
+        UnorderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, shutdown.clone());
+
+    debug!("Starting amqp source, listening to queue {}.", config.queue);
+    let mut consumer = channel
+        .basic_consume(
+            &config.queue,
+            &config.consumer,
+            lapin::options::BasicConsumeOptions::default(),
+            lapin::types::FieldTable::default(),
+        )
+        .await
+        .map_err(|error| {
+            error!(message = "Failed to consume.", error = ?error, internal_log_rate_secs = 10);
```

Comment:
Shouldn't this be an internal event for "Error" per https://github.com/vectordotdev/vector/blob/master/docs/specs/instrumentation.md#error ?

Reply:
This error results in the source not starting up. As discussed on slack we do not need to raise an event in this case.

Reply:
:+1: